File: ttd-dr-ts/src/ttd-dr-agent.ts
/**
 * Test-Time Diffusion Deep Researcher (TTD-DR)
 * Implementation based on arXiv:2507.16075
 */

import { LLMClient } from './llm-client';
import { SearchTool } from './search-tool';
import { Prompts } from './prompts';
import { AgentConfig, AgentState } from './types';
import { Logger } from './utils';

/**
 * Test-Time Diffusion Deep Researcher
 * Implements the complete TTD-DR framework from the paper
 */
export class TTDDRAgent {
  private llm: LLMClient;
  private search: SearchTool;
  private config: AgentConfig;
  private logger: Logger;

  constructor(
    llmClient: LLMClient,
    searchTool: SearchTool,
    config: AgentConfig,
    logger: Logger
  ) {
    this.llm = llmClient;
    this.search = searchTool;
    this.config = config;
    this.logger = logger;
  }

  /**
   * Main research method implementing the full TTD-DR framework
   * Returns: [final_report, agent_state]
   */
  async research(query: string): Promise<[string, AgentState]> {
    this.logger.info(`Starting research for query: ${query}`);

    // Initialize agent state
    const state: AgentState = {
      query,
      plan: '',
      draft: '',
      qaPairs: [],
      revisionHistory: [],
    };

    // Stage 1: Research Plan Generation
    state.plan = await this.stage1GeneratePlan(query);
    this.logger.info('Stage 1 completed: Research plan generated');

    // Generate initial draft (R0 in Algorithm 1)
    state.draft = await this.generateInitialDraft(query, state.plan);
    state.revisionHistory.push(state.draft);
    this.logger.info('Initial draft (R0) generated');

    // Denoising with Retrieval Loop (Algorithm 1)
    for (let step = 1; step <= this.config.maxRevisionSteps; step++) {
      this.logger.info(`\n=== Denoising Step ${step}/${this.config.maxRevisionSteps} ===`);

      // Stage 2a: Generate search question (Algorithm 1, Line 2)
      const question = await this.stage2aGenerateQuestion(state);
      this.logger.info(`Generated question: ${question}`);

      // Stage 2b: Search and synthesize answer (Algorithm 1, Lines 4-5)
      const answer = await this.stage2bSearchAndAnswer(question);
      state.qaPairs.push([question, answer]);
      this.logger.info(`Answer synthesized (${answer.length} chars)`);

      // Denoising: Revise draft with new information (Algorithm 1, Line 6)
      state.draft = await this.denoiseDraft(state);
      state.revisionHistory.push(state.draft);
      this.logger.info(`Draft revised (revision #${state.revisionHistory.length})`);

      // Check if we should stop
      if (step >= this.config.maxRevisionSteps) {
        this.logger.info('Max revision steps reached');
        break;
      }
    }

    // Stage 3: Final Report Generation
    const finalReport = await this.stage3GenerateFinalReport(state);
    this.logger.info('Stage 3 completed: Final report generated');

    return [finalReport, state];
  }

  /**
   * Stage 1: Research Plan Generation (Section 2.1, Stage 1)
   * With optional self-evolution
   */
  private async stage1GeneratePlan(query: string): Promise<string> {
    this.logger.info('Stage 1: Generating research plan...');

    const prompt = Prompts.stage1ResearchPlan(query);

    if (this.config.nPlan === 1 && this.config.sPlan === 0) {
      // Simple generation without self-evolution
      return await this.llm.generate(prompt);
    } else {
      // With self-evolution
      return await this.selfEvolve(
        prompt,
        this.config.nPlan,
        this.config.sPlan,
        'research plan'
      );
    }
  }

  /**
   * Stage 2a: Search Question Generation (Algorithm 1, Line 2)
   * Generate next search query based on current context
   */
  private async stage2aGenerateQuestion(state: AgentState): Promise<string> {
    this.logger.info('Stage 2a: Generating search question...');

    const prompt = Prompts.stage2aSearchQuestion(
      state.query,
      state.plan,
      state.draft,
      state.qaPairs
    );

    let question: string;
    if (this.config.nQuery === 1 && this.config.sQuery === 0) {
      // Simple generation
      question = await this.llm.generate(prompt, 0.8);
    } else {
      // With self-evolution
      question = await this.selfEvolve(
        prompt,
        this.config.nQuery,
        this.config.sQuery,
        'search question'
      );
    }

    return question.trim();
  }

  /**
   * Stage 2b: Answer Searching with RAG (Section 2.1, Stage 2b)
   * Search and synthesize answer from retrieved documents
   */
  private async stage2bSearchAndAnswer(question: string): Promise<string> {
    this.logger.info('Stage 2b: Searching and synthesizing answer...');

    // Perform search
    const searchResults = await this.search.search(question);
    const formattedResults = this.search.formatResults(searchResults);

    // Synthesize answer using RAG
    const prompt = Prompts.stage2bAnswerSynthesis(question, formattedResults);

    if (this.config.nAnswer === 1 && this.config.sAnswer === 0) {
      // Simple synthesis
      return await this.llm.generate(prompt);
    } else {
      // With self-evolution
      return await this.selfEvolve(
        prompt,
        this.config.nAnswer,
        this.config.sAnswer,
        'answer'
      );
    }
  }

  /**
   * Denoising with Retrieval (Algorithm 1, Line 6)
   * Revise draft by incorporating new information
   */
  private async denoiseDraft(state: AgentState): Promise<string> {
    this.logger.info('Denoising: Revising draft with new information...');

    const prompt = Prompts.denoisingRevision(
      state.query,
      state.draft,
      state.qaPairs
    );

    return await this.llm.generate(prompt);
  }

  /**
   * Stage 3: Final Report Generation (Section 2.1, Stage 3)
   * Synthesize all information into final comprehensive report
   */
  private async stage3GenerateFinalReport(state: AgentState): Promise<string> {
    this.logger.info('Stage 3: Generating final report...');

    const prompt = Prompts.stage3FinalReport(
      state.query,
      state.plan,
      state.qaPairs
    );

    if (this.config.nReport === 1 && this.config.sReport === 0) {
      // Simple generation
      return await this.llm.generate(prompt, 0.7, 8192);
    } else {
      // With self-evolution
      return await this.selfEvolve(
        prompt,
        this.config.nReport,
        this.config.sReport,
        'final report'
      );
    }
  }

  /**
   * Generate initial noisy draft (R0 in Algorithm 1)
   * Based on LLM's internal knowledge
   */
  private async generateInitialDraft(query: string, plan: string): Promise<string> {
    this.logger.info('Generating initial draft (R0)...');

    const prompt = Prompts.initialDraft(query, plan);
    return await this.llm.generate(prompt, 0.7, 4096);
  }

  /**
   * Component-wise Self-Evolution (Section 2.2)
   *
   * Algorithm:
   * 1. Generate multiple initial variants
   * 2. For each variant, apply evolution loop:
   *    - Get environmental feedback (LLM-as-a-judge)
   *    - Revise based on feedback
   *    - Repeat for nEvolutionSteps
   * 3. Merge all evolved variants into final output
   */
  private async selfEvolve(
    prompt: string,
    nVariants: number,
    nEvolutionSteps: number,
    contentType: string
  ): Promise<string> {
    this.logger.info(`Self-evolution: ${nVariants} variants, ${nEvolutionSteps} steps`);

    // Step 1: Generate initial variants with diverse parameters
    const variants = await this.llm.generateMultiple(prompt, nVariants, 0.8);

    if (nEvolutionSteps === 0) {
      // No evolution, just merge initial variants
      if (variants.length === 1) {
        return variants[0];
      }
      return await this.mergeVariants(variants, contentType);
    }

    // Steps 2 & 3: Evolution loop for each variant
    const evolvedVariants: string[] = [];

    for (let i = 0; i < variants.length; i++) {
      this.logger.info(`Evolving variant ${i + 1}/${variants.length}...`);
      let current = variants[i];

      for (let step = 0; step < nEvolutionSteps; step++) {
        // Environmental Feedback: Get critique (Section 2.2, Step 2)
        const critiquePrompt = Prompts.selfEvolutionCritique(current, contentType);
        const feedback = await this.llm.generate(critiquePrompt, 0.5);

        // Revision: Improve based on feedback (Section 2.2, Step 3)
        const revisionPrompt = Prompts.selfEvolutionRevision(current, feedback, contentType);
        current = await this.llm.generate(revisionPrompt, 0.7);
      }

      evolvedVariants.push(current);
    }

    // Step 4: Cross-over - Merge evolved variants (Section 2.2, Step 4)
    if (evolvedVariants.length === 1) {
      return evolvedVariants[0];
    }

    return await this.mergeVariants(evolvedVariants, contentType);
  }

  /**
   * Merge multiple variants into single output (Section 2.2, Step 4 - Cross-over)
   */
  private async mergeVariants(variants: string[], contentType: string): Promise<string> {
    this.logger.info(`Merging ${variants.length} variants...`);

    const prompt = Prompts.mergeVariants(variants, contentType);
    return await this.llm.generate(prompt, 0.7, 8192);
  }
}

File: ttd-dr/src/ttd_dr_agent.py
"""
Test-Time Diffusion Deep Researcher (TTD-DR)
Implementation based on arXiv:2507.16075
"""

import logging
from typing import List, Tuple, Dict, Optional
from dataclasses import dataclass
from .llm_client import LLMClient
from .search_tool import SearchTool
from .prompts import Prompts


@dataclass
class AgentConfig:
    """Configuration for TTD-DR agent"""
    max_revision_steps: int = 20
    n_plan: int = 1
    n_query: int = 5
    n_answer: int = 3
    n_report: int = 1
    s_plan: int = 1
    s_query: int = 0
    s_answer: int = 0
    s_report: int = 1
    save_intermediate: bool = True


@dataclass
class AgentState:
    """State maintained throughout the research process"""
    query: str
    plan: str = ""
    draft: str = ""
    qa_pairs: List[Tuple[str, str]] = None
    revision_history: List[str] = None

    def __post_init__(self):
        if self.qa_pairs is None:
            self.qa_pairs = []
        if self.revision_history is None:
            self.revision_history = []


class TTDDRAgent:
    """
    Test-Time Diffusion Deep Researcher
    Implements the complete TTD-DR framework from the paper
    """

    def __init__(
        self,
        llm_client: LLMClient,
        search_tool: SearchTool,
        config: AgentConfig,
        logger: Optional[logging.Logger] = None
    ):
        self.llm = llm_client
        self.search = search_tool
        self.config = config
        self.logger = logger or logging.getLogger(__name__)
        self.prompts = Prompts()

    def research(self, query: str) -> Tuple[str, AgentState]:
        """
        Main research method implementing the full TTD-DR framework
        Returns: (final_report, agent_state)
        """
        self.logger.info(f"Starting research for query: {query}")

        # Initialize agent state
        state = AgentState(query=query)

        # Stage 1: Research Plan Generation
        state.plan = self._stage1_generate_plan(query)
        self.logger.info("Stage 1 completed: Research plan generated")

        # Generate initial draft (R0 in Algorithm 1)
        state.draft = self._generate_initial_draft(query, state.plan)
        state.revision_history.append(state.draft)
        self.logger.info("Initial draft (R0) generated")

        # Denoising with Retrieval Loop (Algorithm 1)
        for step in range(1, self.config.max_revision_steps + 1):
            self.logger.info(f"\n=== Denoising Step {step}/{self.config.max_revision_steps} ===")

            # Stage 2a: Generate search question (Algorithm 1, Line 2)
            question = self._stage2a_generate_question(state)
            self.logger.info(f"Generated question: {question}")

            # Stage 2b: Search and synthesize answer (Algorithm 1, Lines 4-5)
            answer = self._stage2b_search_and_answer(question)
            state.qa_pairs.append((question, answer))
            self.logger.info(f"Answer synthesized ({len(answer)} chars)")

            # Denoising: Revise draft with new information (Algorithm 1, Line 6)
            state.draft = self._denoise_draft(state)
            state.revision_history.append(state.draft)
            self.logger.info(f"Draft revised (revision #{len(state.revision_history)})")

            # Check if we should stop (could implement more sophisticated exit logic)
            if step >= self.config.max_revision_steps:
                self.logger.info("Max revision steps reached")
                break

        # Stage 3: Final Report Generation
        final_report = self._stage3_generate_final_report(state)
        self.logger.info("Stage 3 completed: Final report generated")

        return final_report, state

    def _stage1_generate_plan(self, query: str) -> str:
        """
        Stage 1: Research Plan Generation (Section 2.1, Stage 1)
        With optional self-evolution
        """
        self.logger.info("Stage 1: Generating research plan...")

        prompt = self.prompts.stage1_research_plan(query)

        if self.config.n_plan == 1 and self.config.s_plan == 0:
            # Simple generation without self-evolution
            plan = self.llm.generate(prompt)
        else:
            # With self-evolution
            plan = self._self_evolve(
                prompt=prompt,
                n_variants=self.config.n_plan,
                n_evolution_steps=self.config.s_plan,
                content_type="research plan"
            )

        return plan

    def _stage2a_generate_question(self, state: AgentState) -> str:
        """
        Stage 2a: Search Question Generation (Algorithm 1, Line 2)
        Generate next search query based on current context
        """
        self.logger.info("Stage 2a: Generating search question...")

        prompt = self.prompts.stage2a_search_question(
            query=state.query,
            plan=state.plan,
            draft=state.draft,
            previous_qa=state.qa_pairs
        )

        if self.config.n_query == 1 and self.config.s_query == 0:
            # Simple generation
            question = self.llm.generate(prompt, temperature=0.8)
        else:
            # With self-evolution
            question = self._self_evolve(
                prompt=prompt,
                n_variants=self.config.n_query,
                n_evolution_steps=self.config.s_query,
                content_type="search question"
            )

        return question.strip()

    def _stage2b_search_and_answer(self, question: str) -> str:
        """
        Stage 2b: Answer Searching with RAG (Section 2.1, Stage 2b)
        Search and synthesize answer from retrieved documents
        """
        self.logger.info("Stage 2b: Searching and synthesizing answer...")

        # Perform search
        search_results = self.search.search(question)
        formatted_results = self.search.format_results(search_results)

        # Synthesize answer using RAG
        prompt = self.prompts.stage2b_answer_synthesis(question, formatted_results)

        if self.config.n_answer == 1 and self.config.s_answer == 0:
            # Simple synthesis
            answer = self.llm.generate(prompt)
        else:
            # With self-evolution
            answer = self._self_evolve(
                prompt=prompt,
                n_variants=self.config.n_answer,
                n_evolution_steps=self.config.s_answer,
                content_type="answer"
            )

        return answer

    def _denoise_draft(self, state: AgentState) -> str:
        """
        Denoising with Retrieval (Algorithm 1, Line 6)
        Revise draft by incorporating new information
        """
        self.logger.info("Denoising: Revising draft with new information...")

        prompt = self.prompts.denoising_revision(
            query=state.query,
            current_draft=state.draft,
            qa_pairs=state.qa_pairs
        )

        revised_draft = self.llm.generate(prompt)
        return revised_draft

    def _stage3_generate_final_report(self, state: AgentState) -> str:
        """
        Stage 3: Final Report Generation (Section 2.1, Stage 3)
        Synthesize all information into final comprehensive report
        """
        self.logger.info("Stage 3: Generating final report...")

        prompt = self.prompts.stage3_final_report(
            query=state.query,
            plan=state.plan,
            qa_pairs=state.qa_pairs
        )

        if self.config.n_report == 1 and self.config.s_report == 0:
            # Simple generation
            final_report = self.llm.generate(prompt, max_tokens=8192)
        else:
            # With self-evolution
            final_report = self._self_evolve(
                prompt=prompt,
                n_variants=self.config.n_report,
                n_evolution_steps=self.config.s_report,
                content_type="final report"
            )

        return final_report

    def _generate_initial_draft(self, query: str, plan: str) -> str:
        """
        Generate initial noisy draft (R0 in Algorithm 1)
        Based on LLM's internal knowledge
        """
        self.logger.info("Generating initial draft (R0)...")

        prompt = self.prompts.initial_draft(query, plan)
        draft = self.llm.generate(prompt, max_tokens=4096)

        return draft

    def _self_evolve(
        self,
        prompt: str,
        n_variants: int,
        n_evolution_steps: int,
        content_type: str
    ) -> str:
        """
        Component-wise Self-Evolution (Section 2.2)

        Algorithm:
        1. Generate multiple initial variants
        2. For each variant, apply evolution loop:
           - Get environmental feedback (LLM-as-a-judge)
           - Revise based on feedback
           - Repeat for n_evolution_steps
        3. Merge all evolved variants into final output
        """
        self.logger.info(f"Self-evolution: {n_variants} variants, {n_evolution_steps} steps")

        # Step 1: Generate initial variants with diverse parameters
        variants = self.llm.generate_multiple(prompt, n=n_variants, temperature=0.8)

        if n_evolution_steps == 0:
            # No evolution, just merge initial variants
            if len(variants) == 1:
                return variants[0]
            return self._merge_variants(variants, content_type)

        # Step 2 & 3: Evolution loop for each variant
        evolved_variants = []

        for i, variant in enumerate(variants):
            self.logger.info(f"Evolving variant {i+1}/{len(variants)}...")
            current = variant

            for step in range(n_evolution_steps):
                # Environmental Feedback: Get critique (Section 2.2, Step 2)
                critique_prompt = self.prompts.self_evolution_critique(current, content_type)
                feedback = self.llm.generate(critique_prompt, temperature=0.5)

                # Revision: Improve based on feedback (Section 2.2, Step 3)
                revision_prompt = self.prompts.self_evolution_revision(current, feedback, content_type)
                current = self.llm.generate(revision_prompt, temperature=0.7)

            evolved_variants.append(current)

        # Step 4: Cross-over - Merge evolved variants (Section 2.2, Step 4)
        if len(evolved_variants) == 1:
            return evolved_variants[0]

        final_output = self._merge_variants(evolved_variants, content_type)
        return final_output

    def _merge_variants(self, variants: List[str], content_type: str) -> str:
        """
        Merge multiple variants into single output (Section 2.2, Step 4 - Cross-over)
        """
        self.logger.info(f"Merging {len(variants)} variants...")

        prompt = self.prompts.merge_variants(variants, content_type)
        merged = self.llm.generate(prompt, max_tokens=8192)

        return merged
